core/gather/driver/prepare.js

```diff
@@ -22,17 +22,33 @@ async function enableAsyncStacks(session) {
     await session.sendCommand('Debugger.setAsyncCallStackDepth', {maxDepth: 8});
   };
 
-  // Resume any pauses that make it through `setSkipAllPauses`
-  session.on('Debugger.paused', () => session.sendCommand('Debugger.resume'));
+  /**
+   * Resume any pauses that make it through `setSkipAllPauses`
+   */
+  function onDebuggerPaused() {
+    session.sendCommand('Debugger.resume');
+  }
 
-  // `Debugger.setSkipAllPauses` is reset after every navigation, so retrigger it on main frame navigations.
-  // See https://bugs.chromium.org/p/chromium/issues/detail?id=990945&q=setSkipAllPauses&can=2
-  session.on('Page.frameNavigated', event => {
+  /**
+   * `Debugger.setSkipAllPauses` is reset after every navigation, so retrigger it on main frame navigations.
+   * See https://bugs.chromium.org/p/chromium/issues/detail?id=990945&q=setSkipAllPauses&can=2
+   * @param {LH.Crdp.Page.FrameNavigatedEvent} event
+   */
+  function onFrameNavigated(event) {
     if (event.frame.parentId) return;
     enable().catch(err => log.error('Driver', err));
-  });
+  }
+
+  session.on('Debugger.paused', onDebuggerPaused);
+  session.on('Page.frameNavigated', onFrameNavigated);
 
   await enable();
+
+  return async () => {
```

Comment:
to me this is calling out to be moved onto something with state (e.g. Driver) rather than requiring the caller to hold onto it. I realize it's less of a big deal outside of legacy since it can just be a local, but a general pattern of requiring the enabler to keep around the disabler when we're already keeping a bunch of state for them (in the event listeners, at the protocol enable/disable level, etc) seems excessive. I guess the refactor TODO below may make that irrelevant, though

Reply:
Part of the reason I designed it this way was to reduce code duplication between FR and legacy. I didn't want to create a `new AsyncStackHandler()` that's shared between the two implementations.

I think a combination of moving this logic to the DT log gatherer (which) has state and killing legacy runner once and for all will eliminate the need for this kind of pattern.

Reply:
makes sense 👍